File: japan-go-fe/types/api/responses/grammar.response.ts
import { BaseResponse } from "./base.response";
import { SentenceResponse } from "./common.response";

export interface GrammarResponse extends BaseResponse<number> {
    grammarTitle: string;
    grammarTitleFurigana: string;
    grammarTitleRomaji: string;
    translation: string;
    jlptLevel: number;

    bookTitle: string;
    lessonName: string;
    lessonId: number;

    grammarMeaning: GrammarMeaningResponse;
    grammarStructure: GrammarStructureResponse;
    grammarExample: GrammarExampleResponse;
    grammarNote: GrammarNoteResponse;
}

export interface GrammarExampleResponse extends BaseResponse<number> {
    sentences: SentenceResponse[];
}

export interface GrammarMeaningResponse extends BaseResponse<number> {
    sentences: SentenceResponse[];
}

export interface GrammarNoteResponse extends BaseResponse<number> {
    sentences: SentenceResponse[];
}

export interface GrammarStructureResponse extends BaseResponse<number> {
    sentences: SentenceResponse[];
}

File: japan-go-be/core/service/src/main/java/org/japan/dto/response/grammar/GrammarResponse.java
package org.japan.dto.response.grammar;

import lombok.*;
import lombok.experimental.FieldDefaults;
import lombok.experimental.SuperBuilder;
import org.japan.dto.response.base.BaseResponse;

@SuperBuilder
@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@FieldDefaults(level = AccessLevel.PRIVATE)
public class GrammarResponse extends BaseResponse {
    String grammarTitle;
    String grammarTitleFurigana;
    String grammarTitleRomaji;
    String translation;
    Integer jlptLevel;

    String bookTitle;
    String lessonName;
    Long lessonId;

    GrammarMeaningResponse grammarMeaning;

    GrammarStructureResponse grammarStructure;

    GrammarExampleResponse grammarExample;

    GrammarNoteResponse grammarNote;
}
